retrieval/discovery/azure.go

```diff
@@ -0,0 +1,214 @@
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Debugf("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
+	type target struct {
+		labelSet model.LabelSet
+		err      error
+	}
+
+	ch := make(chan target, len(machines))
+	for i, vm := range machines {
+		go func(i int, vm compute.VirtualMachine) {
+			resourceGroup := strings.ToLower(strings.Split(*vm.ID, "/")[4])
+			labels := model.LabelSet{
+				azureLabelMachineID:            model.LabelValue(*vm.ID),
+				azureLabelMachineName:          model.LabelValue(*vm.Name),
+				azureLabelMachineLocation:      model.LabelValue(*vm.Location),
+				azureLabelMachineResourceGroup: model.LabelValue(resourceGroup),
+			}
+
+			if vm.Tags != nil {
+				for k, v := range *vm.Tags {
+					name := strutil.SanitizeLabelName(k)
+					labels[azureLabelMachineTag+model.LabelName(name)] = model.LabelValue(*v)
+				}
+			}
+
+			// Get the IP address information via seperate call to the network provider.
+			for _, nic := range *vm.Properties.NetworkProfile.NetworkInterfaces {
+				group := strings.Split(*nic.ID, "/")[4]
+				name := strings.Split(*nic.ID, "/")[8]
+				networkInterface, err := client.nic.Get(group, name, "")
+				if err != nil {
+					log.Errorf("Unable to get network interface %s: %s", name, err)
+					tgt := target{labelSet: labels, err: err}
+					ch <- tgt
+					// Get out of this routine because we cannot continue without a network interface.
+					return
+				}
+
+				// Unfortunately Azure does not return information on whether a VM is deallocated.
+				// This information is available via another API call however the Go SDK does not
+				// yet support this. On deallocated machines, this value happens to be nil so it
+				// is a cheap and easy way to determine if a machine is allocated or not.
+				if networkInterface.Properties.Primary == nil {
+					log.Infof("Virtual machine %s looks deallocated. Skipping during Azure SD.", *vm.Name)
+					return
+				}
+
+				if *networkInterface.Properties.Primary {
+					for _, ip := range *networkInterface.Properties.IPConfigurations {
+						if ip.Properties.PrivateIPAddress != nil {
+							labels[azureLabelMachinePrivateIP] = model.LabelValue(*ip.Properties.PrivateIPAddress)
+							address := fmt.Sprintf("%s:%d", *ip.Properties.PrivateIPAddress, ad.port)
+							labels[model.AddressLabel] = model.LabelValue(address)
+						}
+					}
+				}
+
+				tgt := target{labelSet: labels}
+				ch <- tgt
+			}
+		}(i, vm)
+	}
+
+	for range machines {
+		tgt := <-ch
+		if tgt.err != nil {
+			return nil, fmt.Errorf("Unable to complete Azure service discovery: %s", err)
```

Comment:
Errors like this should start with a lowercase letter, as they tend to be concatenated.